File: java/src/test/java/bool/ExplicitTest.java
package bool;

import static org.hamcrest.MatcherAssert.*;
import static org.hamcrest.CoreMatchers.*;
import org.junit.Test;

import java.io.IOException;

import static java.util.Arrays.asList;
import static org.junit.Assert.assertTrue;



public class ExplicitTest {
	@Test
	public void test_and_or_expression() throws IOException {
		Parser parser = new Parser(new Lexer("a && b || c"));
		Expr expr = parser.parseExpr();
		assertThat("a && b || c should explicitly be ((a && b) || c)",
				expr.describeTo(new ExplicitVisitor(),null),
				is ("((a && b) || c)"));
		
	}
	
	@Test
	public void test_or_and_expression() throws IOException {
		Parser parser = new Parser(new Lexer("a || b && c"));
		Expr expr = parser.parseExpr();
		assertThat("a || b && c should explicitly be (a || (b && c))",
				expr.describeTo(new ExplicitVisitor(),null),
				is ("(a || (b && c))"));
	}
}

File: javascript/test/explicit_test.js
var parser = require('../lib').parser;
var Explicit = require('../lib/explicit');
var assert = require('assert');

describe('Explicit', function() {
  it('test and or precedence', function() {
    var expr = parser.parse('@a && @b || @c');
    assert.equal('((@a && @b) || @c)', expr.describeTo(new Explicit(), null));
  });
});

describe('Explicit', function() {
  it('test not precedence', function() {
    var expr = parser.parse('!(@a && @b || !@c)');
    assert.equal('!((@a && @b) || !@c)', expr.describeTo(new Explicit(), null));
  });
});


describe('Explicit', function() {
  it('test or and precedence', function() {
    var expr = parser.parse('@a || @b && @c');
    assert.equal('(@a || (@b && @c))', expr.describeTo(new Explicit(), null));
  });
});


